### ISSUE TITLE
PHPUnit cannot mock PHP 8.1 enums

### ISSUE THREAD

**Issue by hofmannjan1** [sebastianbergmann/phpunit#4791]
PHP 8.1 adds support for enums, which are treated like `final` classes. So, currently, it is not possible to mock enums in PHPUnit. Will this change in the future?

**Comment by sebastianbergmann** [comment]
This will not change: `enum`s cannot be doubled. And I do not think that there is any need to double them.

**Comment by apetitpa** [comment]
Hi @sebastianbergmann,

I'm facing a similar issue in my tests and I think this is related.

`Return value for App\Mandat\Domain\Model\Mandat::getStatus() cannot be generated: Class "App\Mandat\Domain\Enum\StatusEnum" is declared "final" and cannot be doubled.`

It's really weird because it only happens in one test class even if I mock `Mandat` in others.

The only workaround I found yet is to declare `getStatus` return type as nullable `public function getStatus(): ?StatusEnum`

Of course StatusEnum is a PHP 8.1 native backed enum.

`
enum StatusEnum: string
{
    use EnumValuesTrait;

    case MANDATE = 'MANDATE';
    case RECEPTIONNE = 'RECEPTIONNE';
    case EN_COURS_D_EXPERTISE = 'EN_COURS_D_EXPERTISE';
    case A_SIGNER = 'A_SIGNER';

    public static function getValeursStatutsVehiculesAExpertiser(): array
    {
        return [
            self::RECEPTIONNE->value,
            self::EN_COURS_D_EXPERTISE->value,
        ];
    }
}
`

Here is the test class setUp
`
    private StatutEnCoursDExpertiseUpdater $updater;

    public function setUp(): void
    {
        $this->updater = new StatutEnCoursDExpertiseUpdater();
    }
`
Here is the Mandat getStatus function
`
    public function getStatus(): StatusEnum
    {
        return $this->status;
    }
`

And here is the test
`
    public function testUpdateStatutSuccess(): void
    {
        $mandat = $this->createMock(Mandat::class);
        $mandat->method('getStatus')->willReturn(StatusEnum::RECEPTIONNE);

        $mandat->expects($this->once())->method('getStatus');
        $mandat->expects($this->once())->method('setStatus')->with(StatusEnum::EN_COURS_D_EXPERTISE);

        $result = $this->updater->update($mandat);

        $this->assertTrue($result);
    }
`

**Comment by janopae** [comment]
I agree that mocking enums shouldn't be possible, but shouldn't it be possible to mock classes that have methods which return enums? @sebastianbergmann

**Comment by sebastianbergmann** [comment]
Doubling methods that return an `enum` is possible. What is not possible is automatic return value generation when no return value is configured for such a method.

**Comment by janopae** [comment]
@sebastianbergmann Thanks for clarifying this, I now realised that it works when I manually set a return value.

Automatic return value generation uses `0` for ints and `false` for booleans (which are pretty random choices depending on the context). Do you think it would be apropriate to enable automatic return value generation for enums by always returning the first (0th) enum case?

**Comment by sebastianbergmann** [comment]
> Do you think it would be apropriate to enable automatic return value generation for enums by always returning the first (0th) enum case?

No, sorry.
